source/blog/2017-07-12-unprivileged-containers-with-bwrap-oci-and-bubblewrap.md

```diff
@@ -0,0 +1,184 @@
+an user/group mapping in a new user namespace.  Since they run as
+root, they can allow a user to create usernamespaces with multiple
+user mappings.
+This is of course very risky in principle since any user could map an
+uid/gid in the parent namespace to the user namespace gaining full
+control over it.  The system administrator limits the mappings done by
+these tools to what is defined in /etc/subuid and /etc/subgid.  The
+administrator defines a range  of uid/gid that each user can set for
+their user namespaces, of course this is disabled by default.  Each
+line in the file /etc/subuid file specifies the user in the parent
+namespace, the initial uid of the range and the range length.  Same
+thing for /etc/subgid and gids.
+
+For example on my system, for my user I have defined this:
+
+```
+$ id -u
+1000
+
+$ grep ^gscrivano /etc/subgid
+gscrivano:110000:65536
+```
+
+This allows my login process to use “newuidmap” to setup any mapping
+in a new user namespace that points back to my uid 1000 and any other
+defined in the additional range [110000, 110000+65536)
+
+The uid/gid mapping done by bwrap-oci follows these two rules:
+
+The uid/gid defined in the OCI config.json file (i.e. the uid/gid used
+to exec the sandboxed program) must be  mapped to the uid/gid of the
+user in the parent namespace.
+Any other uid/gid in the range [0, AVAILABLE_UIDS] is mapped to the
+additional ids defined for the user in the parent namespace, as
+described above.
+
+At this point we have all we need to run systemd as a non privileged
+user.
+
+In the bwrap-oci repository there is a [demo](https://github.com/projectatomic/bwrap-oci/blob/master/demos/run-systemd/run_demo.sh)
+on how to setup a container, for simplicity I’ll skip the steps
+required to build it here.
+
+Run a container with bwrap-oci is extremely easy once you have the
+rootfs with all the files required by the container and the
+config.json,
+
+I am not root:
+
+```
+$ id -u
+1000
+```
+
+And I do not have any current capabilities:
+
+```
+$ grep ^Cap /proc/self/status
+CapInh:    0000000000000000
+CapPrm:    0000000000000000
+CapEff:    0000000000000000
+CapBnd:    0000003fffffffff
+CapAmb:    0000000000000000
+```
+
+In order to run a container I am required to have a json file
+describing the application to run and a rootfs directory structure
+containing the application:
+
+```
+$ ls -l
+-rw-rw-r--.  1 gscrivano gscrivano   5741 Jul  7 11:53 config.json
+drwxrwxr-x. 18 gscrivano gscrivano   4096 Jul  6 12:26 rootfs
+```
+
+Now I use bwrap-oci to run the container:
+
+```
+$ bwrap-oci run systemd-demo
+systemd 231 running in system mode. (+PAM +AUDIT +SELINUX +IMA -APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN)
+Detected virtualization container-other.
+Detected architecture x86-64.
+
+Welcome to Fedora 25 (Twenty Five)!
+
+....
+```
+
+Thus, systemd running as a non privileged user!
```

Comment:
Worth mentioning this is `systemd --system`; it thinks it's pid 1, versus the already extant `systemd --user`.  I'd say something like "unprivileged init container" ?